### PR TITLE
Add Python 3.12 Support / Remove Python 3.8 Support

### LINKED ISSUE #1065: [SL-1831] [Feature] Add Python 3.12 Support
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing metricflow functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Python 3.12 should be supported since it's the latest version of Python.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

<sub>[SL-1831](https://linear.app/dbt-labs/issue/SL-1831/[feature]-add-python-312-support-remove-python-38-support)</sub>

### PR DESCRIPTION
**Pull request by plypaul**
Resolves #1065 

### Description

This PR updates the code and Github workflows to support Python 3.12.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -7,7 +7,7 @@ name = "metricflow"
 version = "0.205.0"
 description = "Translates a simple metric definition into reusable SQL and executes it against the SQL engine of your choice."
 readme = "README.md"
-requires-python = ">=3.8,<3.12"
+requires-python = ">=3.9,<=3.12"
```
Comment (tlento) [inline pyproject.toml:10]: Python 3.8 is supported through October, so we need to to support 3.8 through the dbt-core release following official EOL.

File: pyproject.toml [hunk 2]
```diff
@@ -33,9 +33,10 @@ dependencies = [
   "graphviz>=0.18.2, <0.21",
   "halo>=0.0.31, <0.1.0",
   "more-itertools>=8.10.0, <10.2.0",
-  "pandas>=1.5.0, <1.6.0",
+  # pandas 2.2 has a bug https://github.com/databricks/databricks-sql-python/issues/334, so using 2.1.
+  "pandas>=2.1, <2.2",
```
Comment (tlento) [inline pyproject.toml:37]: This is super restrictive, I don't think we can block 1.x installs.

Medium term I'd like to just remove pandas as a production dependency altogether, we don't need it for anything.

File: pyproject.toml [hunk 3]
```diff
@@ -33,9 +33,10 @@ dependencies = [
...
   "pydantic>=1.10.0, <1.11.0",
-  "python-dateutil>=2.8.2, <2.9.0",
+  "python-dateutil~=2.9.0",
```
Comment (tlento) [inline pyproject.toml:39]: Please don't use ~= syntax, it breaks PIP resolvers in certain edge cases.

File: pyproject.toml [hunk 4]
```diff
@@ -56,11 +57,11 @@ dev-packages = [
   "pre-commit>=3.2.2, <3.3.0",
   # Bug with mypy: https://github.com/pallets/click/issues/2558#issuecomment-1656546003
   "click>=8.1.6",
-  "pytest-mock>=3.7.0, <3.8.0",
-  "pytest-xdist>=3.2.1, <3.3.0",
-  "pytest>=7.1.1, < 7.2.0",
-  "types-PyYAML",
+  "pytest-mock~=3.12",
+  "pytest-xdist~=3.5",
+  "pytest~=8.0",
```
Comment (tlento) [inline pyproject.toml:62]: No ~= please.

**Review by tlento — CHANGES_REQUESTED** [review]
Per discussion, we probably need to do a bunch of annoying stuff to keep 3.8 through 3.12 here. Sigh.

**Comment by plypaul** [comment]
Closing this out as this was significantly revised, and related changes will be in a another PR.

**Pull request closed without merging**